File: DesignPatternsExercises/java/FlyweightDesignPattern/FlyweightDesignPattern1/Solution/TreeSelector.java
import java.util.HashMap;
import java.util.Map;

public class TreeSelector {
    private final static Map<String, Tree> treeBySpec = new HashMap<>();

    public static Tree getTree(String treeType) {
        if (treeBySpec.containsKey(treeType))
            return treeBySpec.get(treeType);
        else {
            Tree tree = null;
            if (treeType.equals("Oak Tree")) {
                // tree = new __________(8)__________ ();
                tree = new OakTree();
                treeBySpec.put(treeType, tree);
            } else if (treeType.equals("Flowering Cherry")) {
                // tree = new __________(9)__________ ();
                tree = new FloweringCherry();
                treeBySpec.put(treeType, tree);
            } else {
                System.out.println(treeType + " is not found");
            }
            return tree;
        }
    }
}

File: DesignPatternsExercises/ruby/flyweight_design_pattern/flyweight_design_pattern1/solution/tree_selector.rb
require_relative 'oak_tree'
require_relative 'flowering_cherry'

class TreeSelector
  # __________(9)__________ = {}
  @@tree_map = {}

  # def __________(10)__________(tree_type)
  def self.tree(tree_type)
    return @@tree_map.fetch(tree_type) if @@tree_map.key? tree_type

    # tree = __________(11)__________ if tree_type == 'Oak Tree'
    tree = OakTree.new if tree_type == 'Oak Tree'
    # tree = __________(12)__________ if tree_type == 'Flowering Cherry'
    tree = FloweringCherry.new if tree_type == 'Flowering Cherry'
    raise "#{tree_type} is not found" if tree.nil?

    @@tree_map[tree_type] = tree
    tree
  end
end
